### PR TITLE
By the power of javascript, I grant you the latest windows release

### PR DESCRIPTION
**Pull request by nirvn**
@marioba , @m-kuhn , that's a possibility.

### CODE REVIEW

**Comment by nirvn** [comment]
(we could improve the script to detect whether the browser reports linux/windows/etc. to direct to AppImage et al)

**Comment by suricactus** [comment]
Please note that `gh-pages` always has single commit, so it will be replaced every time after build. There is javascript file that can be configured in (material-?)mkdocs for that purpose.
Also, I am not sure if javascript solution is the best way to go, or the same logic during the build process.
Finally, if we go for javascript solution, I would prefer using the `fetch()` API instead.

**Comment by nirvn** [comment]
@suricactus , I always worry about browser compatibility with fetch. But we're in 2022, that's likely not a problem these days. +1.

I'm not sure I fully understand the remaining points. This would just be a static html page doing a JavaScript redirect.

**Comment by suricactus** [comment]
my bad, I got confused which repository I am on. nvm

**Comment by nirvn** [comment]
@suricactus , and in your opinion, we're getting over 90% browser compatibility when using fetch() in 2022? It's definitively a cleaner approach.

**Comment by nirvn** [comment]
@suricactus , fetch variant implemented.

@marioba , I added the same trick for linux, we'd now have https://qfield.org/get_windows and https://qfield.org/get_linux as hyperlinks pointing to the latest {.exe, .AppImage} release.

File: get_linux/index.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<head>
+<title>Get QField for Windows</title>
```
Comment (suricactus) [inline get_linux/index.html:3]: :penguin: :gun:

File: get_linux/index.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+</head>
+<body>
+<script type="text/javascript">
+fetch('https://api.github.com/repos/opengisch/QField/releases/latest')
+.then(response => {
+    if (response.status !== 200) throw new Error();
+    return response.json();
+})
+.then(json => { window.location.replace('https://github.com/opengisch/QField/releases/download/'+json['tag_name']+'/qfield-'+json['tag_name']+'-linux-x64.AppImage') })
+.catch(error => { window.location.replace('https://github.com/opengisch/QField/releases/latest') });
```
Comment (suricactus) [inline get_linux/index.html:13]:
```suggestion
const API_LATEST_RELEASE_URL = 'https://api.github.com/repos/opengisch/QField/releases/latest';
const BROWSER_LATEST_RELEASE_URL = 'https://github.com/opengisch/QField/releases/latest';
const queryParams = new Proxy(new URLSearchParams(window.location.search), {
  get: (searchParams, prop) => searchParams.get(prop),
});

if (queryParams.platform) {
  fetch(API_LATEST_RELEASE_URL)
    .then((response) => {
      if (!response.ok) throw new Error();
      return response.json();
    })
    .then(json => {
      const downloadUrl = `https://github.com/opengisch/QField/releases/download/${json['tag_name']}/qfield-${json['tag_name']}-${queryParams['platform']}`;
      if (json['assets'].find(asset => asset["browser_download_url"] === downloadUrl)) {
        return downloadUrl;
      } else {
        return BROWSER_LATEST_RELEASE_URL;
      }
    })
    .then(browserUrl => {
      window.location.replace(browserUrl);
    });
} else {
  window.location.replace(BROWSER_LATEST_RELEASE_URL);
}
```

Comment (suricactus) [reply get_linux/index.html:13]: Making it like this means we can add whatever links we want at any point of time and reduce code duplication.

File: get_linux/index.html (hunk 1 above)
Comment (suricactus) [reply get_linux/index.html:3]: I will keep it "Get QField" only.

File: get_linux/index.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+</script>
+<p>Redirecting to the windows download page. If this does not work please go to the <a href='https://github.com/opengisch/QField/releases/latest'>latest release page</a>.</p>
```
Comment (suricactus) [inline get_linux/index.html:15]:
```suggestion
<p>Redirecting to the ${queryParams.nickname} download page. If this does not work please go to the <a href='https://github.com/opengisch/QField/releases/latest'>latest release page</a>.</p>
```
We can pass the platform name also as a queryParameter. Needs a bit more js to work.

File: get_linux/index.html (hunk 1 above)
Comment (suricactus) [reply get_linux/index.html:3]: Or make it "Get QField for ${queryParams.nickname}".

File: get_linux/index.html (hunk 3 above)
Comment (nirvn) [reply get_linux/index.html:15]: Actually there I'd just no refer to the platform name.

File: get_linux/index.html (hunk 1 above)
Comment (nirvn) [reply get_linux/index.html:3]: +1 to not name the platform.

File: get_linux/index.html (hunk 2 above)
Comment (nirvn) [reply get_linux/index.html:13]: I'm fine with that alternative approach. What would be the URL? qfield.org/get_package/?platform={windows,linux} ?

Comment (suricactus) [reply get_linux/index.html:13]: Yup, sounds good.

**Comment by nirvn** [comment]
@suricactus , good comments. The PR now implements the following URLs:

https://qfield.org/get_latest/?platform=windows
https://qfield.org/get_latest/?platform=linux

That'll be super useful for the documentation, as well as in communication with people out there through social media / messenger / etc.

**Review by suricactus — APPROVED** [review]
Shall we try it in the wild/

File: get_latest/index.html [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      if (!response.ok) throw new Error();
+      return response.json();
+    })
+    .then(json => {
+      var downloadUrl = "https://github.com/opengisch/QField/releases/download/" + json['tag_name'] + "/qfield-" + json['tag_name'] + "-";
+      if (queryParams['platform'] == "windows") {
+        downloadUrl += "windows-x64.exe";
+      } else if (queryParams['platform'] == "linux") {
+        downloadUrl += "linux-x64.AppImage";
+      }
```
Comment (suricactus) [inline get_latest/index.html:27]: Note with this implementation difference it means you have to edit this file any time you want to download a new platform binary. Why don't you pass the whole suffix of the asset name, as suggested in my original comment?

Comment (nirvn) [reply get_latest/index.html:27]: @suricactus , I thought qfield.org/get_latest/?platform={windows,linux} were nicer links than qfield.org/get_latest/?platform={windows-x64.exe,linux-x64.AppImage}.

I'd rather have to do one edit whenever we add a new platform (which would be once per few years? :) ) than having longer and slightly more "intimidating" URLs.

Comment (suricactus) [reply get_latest/index.html:27]: we experience different laziness levels  :)

Comment (suricactus) [reply get_latest/index.html:27]: A bit too late, but just to satisfy my laziness and your less intimidating URLs, we should have added `else { downloadUrl = queryParams['platform'] ;  }`

**Pull request merged**